Upgrade nudge: Add filter to control block-editor upgrade nudges

sirreal opened the pull request:
#### Changes proposed in this Pull Request:

The new `jetpack_block_editor_enable_upgrade_nudge` filter can be used
to enable or disable the feature.

Fixes https://github.com/Automattic/jetpack/issues/13203

#### Testing instructions:
With a site on the free plan, open the editor and try to add a Simple Payments block.

* By default, no nudges are shown and the block is unavailable.
* Use `add_filter( 'jetpack_block_editor_enable_upgrade_nudge', '__return_true' );` to enable.
* Use `add_filter( 'jetpack_block_editor_enable_upgrade_nudge', '__return_false' );` to disable.

When disabled in development, you'll see a console message like:

> register-jetpack-block.js:65 Block simple-payments couldn't be registered because it is unavailable (missing_plan__nudge_disabled).

When enabled, the upgrade nudge is shown and works as expected.

#### Proposed changelog entry for your changes:

Add filter `jetpack_block_editor_enable_upgrade_nudge` to control upgrade nudges in the block editor.

sirreal committed:
Add filter to control block-editor upgrade nudges

The new 'jetpack_block_editor_enable_upgrade_nudge' filter can be used
to enable or disable the feature.

sirreal committed:
fixup! Add filter to control block-editor upgrade nudges

ockham commented:
Fixes #13203.

ockham commented on class.jetpack-gutenberg.php:
```diff
@@ -188,6 +188,20 @@ public static function set_extension_available( $slug ) {
+		if ( 'missing_plan' === $reason && ! apply_filters( 'jetpack_block_editor_enable_upgrade_nudge', false ) ) {
+			// The block editor applies an upgrade nudge if `missing_plan` is the reason.
+			// Add a suffix to disable and provide informative reason.
+			$reason .= '__nudge_disabled';
```
`'missing_plan' === $reason` is actually not a sufficient, but only a necessary criterion to be displayed: That's why blocks such as VideoPress currently don't display the nudge.

The other necessary criterion is the presence of a `required_plan` item in the `details` object the availability endpoint returns optionally for each unavailable block/plugin. We could thus also piggyback on that criterion.

sirreal replied:
Updated comment in c2f0719d020e3a90894c7bdb030d56806a8006ef

Use _may apply_ instead of _applies_ to indicate that the reason isn't sufficient to ensure the nudge is shown, but changing the reason is significant to ensure that it isn't 🙂

simison reviewed (dismissed):
Tested that this works using local Jetpack site on free plan.

By default nudge is hidden, with filter I can bring it visible.

I haven't looked at code too closely.

jeherve commented on class.jetpack-gutenberg.php:
```diff
@@ -188,6 +188,20 @@ public static function set_extension_available( $slug ) {
+		 *
+		 * @param boolean
+		 */
+		if ( 'missing_plan' === $reason && ! apply_filters( 'jetpack_block_editor_enable_upgrade_nudge', false ) ) {
```
I like this, but I would move the docblock directly with the filter, by splitting the condition onto 2 different lines. I think it may be a bit more readable?

```php
if (
	// Here we could even have a comment explaining this: https://github.com/Automattic/jetpack/pull/13228/files#r314138211
	'missing_plan' === $reason
	/**
	 * Filter 'jetpack_block_editor_enable_upgrade_nudge' with `true` to enable or `false`
	 * to disable paid feature upgrade nudges in the block editor.
	 *
	 * @since 7.7.0
	 *
	 * @param boolean
	 */
	&& ! apply_filters( 'jetpack_block_editor_enable_upgrade_nudge', false )
) {
```

What do you think?

sirreal replied:
👍 Good idea. Applied in 55b075a7976b7d0b5c5f955a284a0c369cf2082e

sirreal committed:
Update code structure and comments for clarity

sirreal committed:
Updated comment for accuracy

There are more conditions involved, opted for "may apply" instead of
"applies."

jeherve reviewed (approved):
This works well for me. 👍

sirreal commented:
r195630-wpcom
